File: Problem Patterns/Phase1 - Foundation/Python/15.3Sum.py
#!/bin/env python3


class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        threesum_combinations = []
        nums_sorted = sorted(nums)
        for ptr1 in range(len(nums_sorted) - 2):
            ptr2 = ptr1 + 1
            ptr3 = len(nums_sorted) - 1
            if ptr1 > 0 and nums_sorted[ptr1] == nums_sorted[ptr1 - 1]:
                continue
            while ptr3 > ptr2:
                while (
                    ptr3 > ptr2
                    and nums_sorted[ptr1] + nums_sorted[ptr2] + nums_sorted[ptr3] != 0
                ):
                    while (
                        ptr3 > ptr2
                        and nums_sorted[ptr1] + nums_sorted[ptr2] + nums_sorted[ptr3]
                        > 0
                    ):
                        ptr3 -= 1
                    while (
                        ptr3 > ptr2
                        and nums_sorted[ptr1] + nums_sorted[ptr2] + nums_sorted[ptr3]
                        < 0
                    ):
                        ptr2 += 1
                if (
                    ptr3 > ptr2
                    and nums_sorted[ptr1] + nums_sorted[ptr2] + nums_sorted[ptr3] == 0
                ):
                    threesum_combinations.append(
                        [nums_sorted[ptr1], nums_sorted[ptr2], nums_sorted[ptr3]]
                    )
                ptr3 -= 1
                while ptr3 > ptr2 and nums_sorted[ptr3] == nums_sorted[ptr3 + 1]:
                    ptr3 -= 1
                ptr2 += 1
                while ptr3 > ptr2 and nums_sorted[ptr2] == nums_sorted[ptr2 - 1]:
                    ptr2 += 1
        return threesum_combinations


if __name__ == "__main__":
    sol = Solution()
    print(sol.threeSum([-1, 0, 1, 2, -1, -4]))
    print(sol.threeSum([0, 1, 1]))
    print(sol.threeSum([0, 0, 0]))
    print(sol.threeSum([-4, -1, -1, 0, 1, 2]))
    print(sol.threeSum([-2, -2, 0, 2, 2]))

File: Problem Patterns/Phase1 - Foundation/Rust/15_3Sum.rs
#[derive(Debug)]
struct Solution;

impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums: Vec<i32> = nums;
        nums.sort_unstable();
        let mut ptr1: usize;
        let mut three_sum_combinations: Vec<Vec<i32>> = vec![];
        let mut ptr2;
        let mut ptr3;
        for ptr1 in 0..(nums.len() - 2) {
            ptr2 = ptr1 + 1;
            ptr3 = nums.len() - 1;
            if ptr1 > 0 && nums[ptr1] == nums[ptr1 - 1] {
                continue;
            }
            while ptr3 > ptr2 {
                while ptr3 > ptr2 && nums[ptr1] + nums[ptr2] + nums[ptr3] != 0 {
                    while ptr3 > ptr2 && nums[ptr1] + nums[ptr2] + nums[ptr3] < 0 {
                        ptr2 += 1;
                    }
                    while ptr3 > ptr2 && nums[ptr1] + nums[ptr2] + nums[ptr3] > 0 {
                        ptr3 -= 1;
                    }
                }
                if ptr3 > ptr2 && nums[ptr1] + nums[ptr2] + nums[ptr3] == 0 {
                    three_sum_combinations.push(vec![nums[ptr1], nums[ptr2], nums[ptr3]]);
                }
                ptr2 += 1;
                while ptr3 > ptr2 && nums[ptr2] == nums[ptr2 - 1] {
                    ptr2 += 1;
                }
                ptr3 -= 1;
                while ptr3 > ptr2 && nums[ptr3] == nums[ptr3 + 1] {
                    ptr3 -= 1;
                }
            }
        }
        three_sum_combinations
    }
}

fn main() {
    println!("{:?}", Solution::three_sum(vec![-1, 0, 1, 2, -1, -4]));
    println!("{:?}", Solution::three_sum(vec![0, 1, 1]));
    println!("{:?}", Solution::three_sum(vec![0, 0, 0]));
}
